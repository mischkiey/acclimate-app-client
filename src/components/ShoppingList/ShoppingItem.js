import React from 'react';

function ShoppingItem(props) {
    return (
        <>
            <div className='list-group'>
                <button
                    className='y-btn'
                    onClick={(e) => props.handleCheckShoppingItem(e, props.user_id, props.user_shopping_item_id, props.user_shopping_item_completed)}
                >
                    <i className="material-icons">check</i>
                </button>
                <input
                    className=
                        {
                            (props.user_shopping_item_completed)
                                ? 'completed item-input' 
                                : 'item-input'
                        }
                    type='text'
                    defaultValue={props.user_shopping_item}
                    onChange={(e) => {props.handleEditShoppingItem(e, props.user_id, props.user_shopping_item_id)}}
                />
                <button
                    className='r-btn'
                    onClick={() => props.handleDeleteShoppingItem(props.user_shopping_item_id)}
                >
                    <i className="material-icons">delete</i>
                </button>
            </div>
        </>
    );
};

export default ShoppingItem;