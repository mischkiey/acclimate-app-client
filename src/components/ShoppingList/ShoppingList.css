.add-shopping-item-form-group {
    display: flex;
    flex-flow: row;
}