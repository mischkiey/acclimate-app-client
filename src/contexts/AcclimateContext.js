import React from 'react';

const AcclimateContext = React.createContext({});

export default AcclimateContext;