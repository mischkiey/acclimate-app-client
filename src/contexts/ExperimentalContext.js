import React, { Component } from 'react';
import TokenService from '../services/token-services';

export const ExperimentalContext = React.createContext({});

export class ExperimentalProvider extends Component {
    state = {
        log: false,
        tasks: [],
        shoppingItems: [],
    };

    handleUserLog = () => {
        const log = TokenService.hasAuthToken();
        this.setState({log});
    };

    handleUserLogOut = () => {
        TokenService.clearAuthToken();
        this.handleUserLog();
    };

    setTasks = (tasks) => {
        this.setState({tasks})
    };

    setShoppingItems = (shoppingItems) => {
        this.setState({shoppingItems})
    };

    render() {
        const value = {
            log: this.state.log,
            tasks: this.state.tasks,
            shoppingItems: this.state.shoppingItems,

            handleUserLog: this.handleUserLog,
            handleUserLogOut: this.handleUserLogOut,
            setTasks: this.setTasks,
            setShoppingItems: this.setShoppingItems,
        };

        return (
            <ExperimentalContext.Provider value={value}>
                {this.props.children}
            </ExperimentalContext.Provider>
        );
    };
};

