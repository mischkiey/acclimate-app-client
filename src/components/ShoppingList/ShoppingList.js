import React, { Component } from 'react';
import APIService from '../../services/api-services';
import TokenService from '../../services/token-services';
import ShoppingItem from './ShoppingItem';
import './ShoppingList.css';

class ShoppingList extends Component {
    state = {
        items: [],
        newUserShoppingItem: {},
        error: null,
    };
    
    handleAddShoppingItem = (e) => {
        e.preventDefault();

        this.setState({error: null})

        const user_shopping_item = e.target.add_user_shopping_item;
        const newUserShoppingItem = {
            user_shopping_item: user_shopping_item.value,
        };
        const token = TokenService.getAuthToken();

        return APIService.post('/disaster/user/shopping', newUserShoppingItem, token)
            .then(response => {
                this.state.items.push(response);
                const newShoppingItems = this.state.items;
                this.setState({shoppings: newShoppingItems});
                user_shopping_item.value = '';
            })
            .catch(error => {
                this.setState({...error});
            });
    };

    handleEditShoppingItem = (e, user_id, user_shopping_item_id) => {
        this.setState({error: null});

        const newUserShoppingItem = {
            user_shopping_item_id,
            user_shopping_item: e.target.value,
        };
        const token = TokenService.getAuthToken();

        return APIService.patch(`/disaster/user/shopping/${user_shopping_item_id}`, newUserShoppingItem, token)
            .then(() => {
                const newUserShoppingItems = this.state.items.filter(shopping => shopping.user_shopping_item_id !== newUserShoppingItem.user_shopping_item_id);
                newUserShoppingItem.user_id = user_id;
                newUserShoppingItems.push(newUserShoppingItem);
                newUserShoppingItems.sort((a, b) => {
                    if(a.user_shopping_item_id < b.user_shopping_item_id) {
                        return - 1;
                    } else {
                        return 1;
                    };
                });
                this.setState({items: newUserShoppingItems})
            })
            .catch(error => {
                this.setState({...error});
            });
    };

    handleDeleteShoppingItem = (user_shopping_item_id) => {
        this.setState({error: null})

        const token = TokenService.getAuthToken();
        return APIService.del(`/disaster/user/shopping/${user_shopping_item_id}`, token)
            .then(response => {
                const newUserShoppingItems = this.state.items.filter(shopping => shopping.user_shopping_item_id !== user_shopping_item_id);
                this.setState({items: newUserShoppingItems});
            })
            .catch(error => {
                this.setState({error: null})
            })
    }

    componentDidMount() {
        const token = TokenService.getAuthToken();
        return APIService.get('/disaster/user/shopping', token)
            .then(items => {
                this.setState({items});
            })
            .catch(error => {
                this.setState({...error});
            });
    };

    render () {
        const items = this.state.items.map(shopping => <ShoppingItem key={shopping.user_shopping_item_id} handleDeleteShoppingItem={this.handleDeleteShoppingItem} handleEditShoppingItem={this.handleEditShoppingItem} {...shopping} />)

        return (
            <>
                <h2 className='center'>Shopping List</h2>
                <form className='list-group' onSubmit={(e) => this.handleAddShoppingItem(e)} >
                    <input className='item-input' name='add_user_shopping_item' placeholder='Enter shopping item' type='text' />
                    <button className='y-btn'><i className="material-icons">add_circle</i></button>
                </form>
                {items}
            </>
        );
    };
};

export default ShoppingList;