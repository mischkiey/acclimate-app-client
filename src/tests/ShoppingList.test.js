import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
import { AcclimateProvider } from '../contexts/AcclimateContext';

import ShoppingList from '../components/ShoppingList/ShoppingList';

describe(`ShoppingList Component`, () => {
    describe(`Smoke test`, () => {
        it(`Renders without crashing`, () => {
            const div = document.createElement('div');
            ReactDOM.render(
                <BrowserRouter> 
                    <AcclimateProvider>
                        <ShoppingList />
                    </AcclimateProvider>  
                </BrowserRouter>
                , div);
            ReactDOM.unmountComponentAtNode(div);
        });
    });
    
    describe(`Snapshot test`, () => {
        it(`Renders the UI as expected`, () => {
            const tree = renderer
                .create(
                    <BrowserRouter> 
                        <AcclimateProvider>
                            <ShoppingList />
                        </AcclimateProvider>  
                    </BrowserRouter>
                )
                .toJSON()
            expect(tree).toMatchSnapshot()
        });
    });
})