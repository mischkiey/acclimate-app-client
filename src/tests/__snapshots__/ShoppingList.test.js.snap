// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShoppingList Component Snapshot test Renders the UI as expected 1`] = `
Array [
  <h2
    className="center"
  >
    Shopping List
  </h2>,
  <form
    className="list-group"
    onSubmit={[Function]}
  >
    <input
      className="item-input"
      name="add_user_shopping_item"
      placeholder="Enter shopping item"
      type="text"
    />
    <button
      className="tooltip y-btn"
    >
      <i
        className="material-icons"
      >
        add_circle
      </i>
      <span
        className="long tooltiptext"
      >
        Click to add item to your shopping list
      </span>
    </button>
  </form>,
]
`;
